gustaf/ffd.py

```diff
@@ -1,4 +1,571 @@
+            
+            # if not self._spline.control_point_bounds == self._mesh.bounds:
+            #     self._logw("The bounds of the control "
+            #                "spline are not the same.")
+
+
+    @property
+    def mesh(self,) -> MESH_TYPES:
+        """Returns copy of current mesh. Before copying, it applies 
+        deformation.
+
+        Returns
+        -------
+        MESH_TYPES
+            Current Mesh with the deformation according to the current spline.
+        """        
+        self._deform()
+        return self._mesh.copy()
+
+    @mesh.setter
+    def mesh(self, mesh: MESH_TYPES):
+        """
+        Sets mesh. If it is first time, the copy of it will be saved as 
+        original mesh. If spline is already defined and in transformed status, 
+        it applies transformation directly.
+
+        Parameters
+        -----------
+        mesh: MESH_TYPES
+            Mesh used for the FFD
+
+        Returns
+        --------
+        None
+        """
+        if self._mesh is not None:
+            self._logw("Resetting original Mesh, this is not intended "
+            "behavior. Please create a new FFD object when replacing the mesh"
+            ".")
+
+        self._logi("Setting mesh.")
+        self._logi("Mesh Info:")
+        self._logi(
+            "  Vertices: {v}.".format(v=mesh.vertices.shape)
+        )
+        self._logi(
+            "  Bounds: {b}.".format(b=mesh.get_bounds())
+        )
+        self._mesh = mesh.copy() # copy to make sure given mesh stays untouched
+
+        self._scale_mesh_vertices()
+        self.calculated = False
+
+    @property
+    def spline(self):
+        """
+        Returns a copy of current spline. This is to prevent unnecessary
+        tangling
+
+        Parameters
+        -----------
+        None
+
+        Returns
+        --------
+        self._spline: Spline 
+        """
+        self._logd("Returning copy of current spline.")
+        return self._spline.copy() if self._spline is not None else None
+
+    @spline.setter
+    def spline(self, spline: SPLINE_TYPES):
+        """
+        Sets spline. The spline parametric range bounds will be converted 
+        into the bounds [0,1]^para_dim.
+
+        Parameters
+        -----------
+        spline: SPLINE_TYPES
+            New Spline for the next deformation
+
+        Returns
+        --------
+        None
+        """
+        self._spline = spline.copy()
+        self._scale_parametric_dimension_to_hypercube()
+
+        self._logi("Setting Spline.")
+        self._logi("Spline Info:")
+        self._logi(
+            f"  Parametric dimensions: {spline.para_dim}."
+        )
+        self.calculated = False
+
+    @property
+    def is_partial(self):
+        """
+        Returns is_partial.
+
+        Parameters
+        -----------
+        None
+
+        Returns
+        --------
+        self._is_partial: bool
+        """
+        if self._is_partial:
```

Comment:
You could always throw a NotImplemented-Exception if is_partial is set true. This would be more consistent